api/src/utils/transaction.ts

```diff
@@ -7,10 +9,47 @@ import type { Knex } from 'knex';
  * Can be used to ensure the handler is run within a transaction,
  * while preventing nested transactions.
  */
-export const transaction = <T = unknown>(knex: Knex, handler: (knex: Knex) => Promise<T>): Promise<T> => {
+export const transaction = async <T = unknown>(knex: Knex, handler: (knex: Knex) => Promise<T>): Promise<T> => {
 	if (knex.isTransaction) {
 		return handler(knex);
 	} else {
-		return knex.transaction((trx) => handler(trx));
+		try {
+			return await knex.transaction((trx) => handler(trx));
+		} catch (error: any) {
+			const client = getDatabaseClient(knex);
+
+			/**
+			 * This error code indicates that the transaction failed due to another
+			 * concurrent or recent transaction attempting to write to the same data.
+			 * This can usually be solved by restarting the transaction on client-side
+			 * after a short delay, so that it is executed against the latest state.
+			 *
+			 * @link https://www.cockroachlabs.com/docs/stable/transaction-retry-error-reference
+			 */
+			const COCKROACH_RETRY_ERROR_CODE = '40001';
```

Comment:
Should this be moved out of the function to the start of this file?

Reply:
Why?

Reply:
Totally based on a feeling, call it personal preference. 
It feels like an important constant, especially with it being documented that well, that might as well be moved to the module level. Which is weird considering I'm fine with `MAX_ATTEMPTS` and `BASE_DELAY` being there. But those feel less important and I found a few occurrences in our code base where we actually have all-caps constants defined in functions.

I'm fine with it either way tho :)

Reply:
I see, thanks 👍 
Works either way for me too :) I think in this case, I tend a little more towards the current way, so that the variable is close to the context where it is actually used (which is only in case of error).